### ISSUE TITLE
[BUG] Powershell code executed every time after allocating an Azure VM

### ISSUE THREAD

**Issue by mdurini** [Azure/azure-libraries-for-net#1159]
**Describe the bug**
we noticed that each PowerShell script "injected" via our internal application built using Azure Management Fluent API 1.34.0 is re-executed every time I allocate the VM

**Setup (please complete the following information):**
 - IDE : Visual Studio 2019
 - Version of the Library used: Azure Management Fluent API 1.34.0

**Expected behavior**
I don't expect to have the script executed every time 

**Additional context**
Every time I start a machine, the previously executed PowerShell script is triggered again. The script is stored in a storage account and the application is taking care to download and execute it inside the VM. I saw that the file present every time inside the folder "C:\Packages\Plugins\Microsoft.CPlat.Core.RunCommandWindows\1.1.5\Downloads\": I don't have an extension connected to the VM that can justify this weird behavior.

**Comment by yungezz** [comment]
hi @mdurini thanks for reporting the issue. @xccc-msft could you pls help to have a quick look is it due to extension init script?

**Comment by xccc-msft** [comment]
@mdurini Could you please share some code snippet how you created the vm? Thanks.

**Comment by mdurini** [comment]
here you can find the template and the parameter files used to create the vm from our internal  image gallery 

```json
{
  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "adminUsername": {
      "type": "string",
      "minLength": 1,
      "metadata": {
        "description": "Username for the Virtual Machine."
      }
    },
    "adminPassword": {
      "type": "securestring",
      "metadata": {
        "description": "Password for the Virtual Machine."
      }
    },
    "Customername": {
      "type": "string",
      "minLength": 1,
      "metadata": {
        "description": "Customer name"
      }
    },

    "CustomerEnv": {
      "type": "string",
      "minLength": 1,
      "metadata": {
        "description": "Customer Environment identifier: type c1 for production, s1 for sandbox, d1 for developmnet"
      }
    },

    "virtualmachinesize": {
      "type": "string",
      "minLength": 1,
      "metadata": {
        "description": "Virtual machine size"
      }
    },


    "ImageResourceID": {
      "type": "string",
      "minLength": 1,
      "metadata": {
        "description": "Image Resource ID"
      }
    }
       
  
  },
  "variables": {
    "virtualNetworkName": "[concat(parameters('Customername'), '-vn')]",
    "subnetName": "[concat(parameters('Customername'), '-sn1')]",
    "StorageType": "Premium_LRS",
    "publicIPAddressType": "Dynamic",
    "secgroupName": "[concat(parameters('Customername'), '-sg')]",
    "vhdStorageContainerName": "vhds",
    "vmName": "[concat(parameters('Customername'),'-',parameters('CustomerEnv'))]",
    "OSDiskName": "[concat(variables('vmName'), '-dsk.vhd')]",
    "nicName": "[concat(variables('vmName'), '-eth')]",
    "publicIPAddressName": "[concat(variables('vmName'), '-ip')]",
    "vmSize": "[parameters('virtualmachinesize')]",
    "vnetId": "[resourceId('Microsoft.Network/virtualNetworks', variables('virtualNetworkName'))]",
    "subnetRef": "[concat(variables('vnetId'), '/subnets/', variables('subnetName'))]",
    "vhdStorageName": "[concat(parameters('Customername'), 'strg01')]",
    "computername": "[concat(replace(substring(padLeft(parameters('Customername'),7,'*'),0,7),'*',''),'-',parameters('CustomerEnv'))]",
    "diskName": "[concat(variables('vmName'), '-dsk0')]"
  },
  "resources": [
    {
      "apiVersion": "2015-06-15",
      "type": "Microsoft.Network/publicIPAddresses",
      "name": "[variables('publicIPAddressName')]",
      "location": "[resourceGroup().location]",
      "tags": {
        "displayName": "PublicIPAddress",
        "CustomerName": "[parameters('Customername')]"
      },
      "properties": {
        "publicIPAllocationMethod": "[variables('publicIPAddressType')]",
        "dnsSettings": {
          "domainNameLabel": "[variables('vmName')]"
        }
      }
    },
    {
      "apiVersion": "2015-06-15",
      "type": "Microsoft.Network/networkInterfaces",
      "name": "[variables('nicName')]",
      "location": "[resourceGroup().location]",
      "tags": {
        "displayName": "NetworkInterface",
        "CustomerName": "[parameters('Customername')]"
      },
      "dependsOn": [
        "[concat('Microsoft.Network/publicIPAddresses/', variables('publicIPAddressName'))]"
      ],
      "properties": {
        "ipConfigurations": [
          {
            "name": "ipconfig1",
            "properties": {
              "privateIPAllocationMethod": "Dynamic",
              "publicIPAddress": {
                "id": "[resourceId('Microsoft.Network/publicIPAddresses', variables('publicIPAddressName'))]"
              },
              "subnet": {
                "id": "[variables('subnetRef')]"
              }
            }
          }
        ],
        "networkSecurityGroup": {
          "id": "[resourceId('Microsoft.Network/networkSecurityGroups', variables('secgroupName'))]"
        }
      }
    },

    {
      "apiVersion": "2018-04-01",
      "type": "Microsoft.Compute/virtualMachines",
      "name": "[variables('vmName')]",
      "location": "[resourceGroup().location]",
      "tags": {
        "displayName": "VirtualMachine"
      },
      "dependsOn": [
        "[concat('Microsoft.Network/networkInterfaces/', variables('nicName'))]"
      ],
      "properties": {
        "hardwareProfile": {
          "vmSize": "[variables('vmSize')]"
        },

        "osProfile": {
          "computerName": "[variables('computername')]",
          "adminUsername": "[parameters('adminUsername')]",
          "adminPassword": "[parameters('adminPassword')]",
          "windowsConfiguration": {
            "provisionVMAgent": true,
            "enableAutomaticUpdates": true
          }
        },
        "storageProfile": {
          "osDisk": {
            "name": "[variables('diskName')]",
            "createOption": "FromImage",
            "managedDisk": {
              "storageAccountType": "[variables('StorageType')]"
            }
          },
          "imageReference": {
            "id": "[parameters('ImageResourceID')]"
          }
        },




        "networkProfile": {
          "networkInterfaces": [
            {
              "id": "[resourceId('Microsoft.Network/networkInterfaces', variables('nicName'))]"
            }
          ]
        }




      },
      "resources": [
        {
          "apiVersion": "2015-06-15",
          "type": "Microsoft.Compute/virtualMachines/extensions",
          "name": "[concat(variables('vmName'),'/BGInfo')]",
          "location": "[resourceGroup().location]",
          "dependsOn": [
            "[concat('Microsoft.Compute/virtualMachines/', variables('vmName'))]"
          ],
          "tags": {
            "displayName": "BgInfo"
          },
          "properties": {
            "publisher": "Microsoft.Compute",
            "type": "BGInfo",
            "typeHandlerVersion": "2.1",
            "autoUpgradeMinorVersion": true
          }
        }
      ]
    }
    ]
}
```

```json
{
  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
  "contentVersion": "1.0.0.0",
  "parameters": {
    "adminUsername": {
      "value": "adminvm"
    },
    "Customername": {
      "value": "customer03"
    },
    "CustomerEnv": {
      "value": "c1"
    },
    "virtualmachinesize": {
      "value": "Standard_E2s_v3"
    },
    "adminPassword": {
      "value": "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
    },
    "ImageResourceID": {
      "value": "/subscriptions/XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX/resourceGroups/XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX/providers/Microsoft.Compute/galleries/XXXXXXXXXXXXXXXXX/images/Image-10-6-0/versions/XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
    }
  }
}```

**Comment by xccc-msft** [comment]
@mdurini Thanks for sharing. The extension `BgInfo` runs every boot while it seems that would not execute the script. I just add label of `Service Attention` so the service team could help check your issue. Thanks.

**Comment by mdurini** [comment]
Thank you @xccc-msft . do you need something else for future investigation? Like for example the vs code used to execute the PowerShell script or some logs from the agent inside the VM. I noticed that the code is executed also after restoring the OS disk with a newest one created from a snapshot captured before executing the script the first time. Another interesting thing is that the script downloaded and execute is the same version of the first run (I tried to upload a newest version between the two runs in the storage account)

**Comment by Drewm3** [comment]
@axayjo, could you look into this issue which looks to be around the RunCommand?

**Comment by mdurini** [comment]
Hello, any news on that topic?

**Comment by mdurini** [comment]
@Drewm3  can you please escalate this issue?

**Comment by koralski** [comment]
The deployment template shows you are using custom image. Most probably when you prepared that image it was from existing VM and previously executed a RunCommand action on the VM which didn't complete. Image was prepared from the VM in a state that script is not marked as completed which explain why it re-execute every time you create a VM from that image.
I can investigate a live VM if you send me the resource Id to atanask@microsoft.com

**Comment by mdurini** [comment]
Hello @koralski, i've opened a ticket to the azure support after doing some tests: let me recap them

- After reallocating the machine without swapping the OS Disk, some files were updated inside the folder C:\Packages\Plugins\Microsoft.CPlat.Core.RunCommandWindows\1.1.5\” but the script seems that was not executed
- The script is executed again only after swapping the OS disk with a snapshot of the machine captured before executing the RunCommand
- The problem is occurring using the APIs and the "Run command" option from the Azure Portal
- The problem is present with a custom template and a standard image from Market Place

**Comment by mdurini** [comment]
After discussing and investigating with @koralski , he provided me the following solution

> Thank you for providing detailed logs for your scenario. We investigated why run command got executed when you swap the OS disk.
Root cause is that VM has been previously used to call RunCommand API and this remains in the VM model we store internally. This model does not change when you swap the disk and the VM configuration stull contains (a hidden for you) RunCommandWindows extension which gets executed on the new OSDisk because the flag if previously executed is stored in the registry (which is gone when you swap the disk). We agree that fix is needed for your scenario that will cleanup the runcommand from the VM model as well when OSdisk get swapped.
Until we prepare and deploy the fix (usually within 1 month cycle) I can suggest a mitigation that you can apply just before you swap the disks.
You can remove the run command extension from the VM model by calling a special run command with commandId = RemoveRunCommandWindowsExtension
It is not available on the portal but you can use REST, CLI, Powershell or .Net API. For example:
Invoke-AzureRmVMRunCommand -ResourceGroupName 'rgname' -Name 'vmname' -CommandId 'RemoveRunCommandWindowsExtension'

```
Powershell
Invoke-AzureRmVMRunCommand -ResourceGroupName 'rgname' -Name 'vmname' -CommandId 'RemoveRunCommandWindowsExtension'
```

```
CLI
az vm run-command invoke -g MyResourceGroup -n MyVM  --command-id RemoveRunCommandWindowsExtension 
```

> This will delete the extension and after you swap the disk it will not be installed and run again.

**Comment by axayjo** [comment]
@mdurini Did the solution work for you? Can we close the issue?

**Comment by mdurini** [comment]
@axayjo yes it worked. do you know if the fix mentioned by @koralski is already in place?

**Comment by axayjo** [comment]
@koralski can you confirm if the fix has been rolled out yet? @mdurini I'll close the issue once the code has been rolled out.

**Comment by koralski** [comment]
No fixed yet since its a low priority for this corner case.

**Comment by axayjo** [comment]
@mdurini, since this is a low priority bug, the fix won't be in for a while. I've added this issue to the backlog on my end. Do you want to keep this issue open or can I go ahead and close it?

**Comment by mdurini** [comment]
@axayjo how could I be notified when the fix will be deployed if the issue will be closed?

**Comment by mdurini** [comment]
Do we have an ETA for the fix?

**Comment by axayjo** [comment]
@mdurini, We don't have an ETA yet for the fix. This is a low priority bug and will not be fixed in the immediate term. Please feel free to reopen the issue in case the given workaround doesn't work.
